zeth-contracts/contracts/BaseMixer.sol

```diff
@@ -150,36 +163,110 @@ contract BaseMixer is MerkleTreeMiMC7, ERC223ReceivingContract {
+        // ------- now the first sum corresponds to encoding end_byte and the second start_byte
+        //         we note that res_bin[i] = start_byte[i] for i < 8
+        //         likewise, res_bin[i] = end_byte[i - 8] for 7 < i < 16
+        //     = Sum_{i=0}^{8-1} start_byte[i] * 2**(16-1-i) + Sum_{i=8}^{16-1} end_byte[i-8] * 2**(16-1-i)
+        // ------- we reorder the indices to have end_byte[i] (we have j = i - 8), and factorise by 2**8 in the left sum
+        //     = 2**8 * Sum_{i=0}^{8-1} end_byte[i] * 2**(8-1-i) + Sum_{j=0}^{8-1} start_byte[j] * 2**(8-1-j)
+        // ------- we can now see the relationship between the encoding of res_bin and the encodings of end_byte and start_byte
+        //     = 2**8 * uint8(start_byte) + uint8(end_byte)
+        uint16 res = (2**8) * uint8(start_byte) + uint8(end_byte);
+        // The bit representation of res is now something like this (b_x representing the x^{th} bit):
+        // b_0 || ... || b_{start-1} || b_{start} || ... || b_{end} || b_{end + 1} || ... || b_15
+        //                              ^                         ^
+        //                              |_________________________|
+        //                                           |
+        //                                  (length) asked bits
+        // To get the needed bits, we need to discard the unneeded ones before and after.
+        // First we compute the index of the first bit needed (bit_start) in start_byte.
+        uint index_bit_start = (padding_start + offset + start) % 8;
+        // We need to remove all the bits before this index.
+        // As the most significant bit of the byte is the first, the binary operation
+        // "res << index_bit_start" corresponds to multiplying by 2**index_bit_start.
+        res = res * uint16(2**index_bit_start);
+        // We now have something like this:
+        // b_{start} || ... || b_{end} || b_{end + 1} || ... || b_15 || 0 || ... || 0
+        // ^                         ^                                  ^           ^
+        // |_________________________|                                  |___________|
+        //              |                                                     |
+        //     (length) asked bits                                          padding
+        // We need to remove all the bits after b_end which is at location length-1, hence discard (16 - length) bits on the right.
+        // Similarly, the binary operation "res >> 16-length" corresponds to dividing by 2**(16 - length)
         res = res / uint16(2**(16 - length));
 
+        // We now have something like this:
+        // 0 || ... || 0 || b_{start} || ... || b_{end}
+        //                  ^                         ^
+        //                  |_________________________|
+        //                               |
+        //                      (length) asked bits
+        // All the asked bits are in the second byte, we thus return this one.
```

Comment:
> All the asked bits are in the second byte, we thus return this one.

This is only true if `length` is smaller than a byte, and it is just required to be < 16, so it can be more than a byte

Reply:
That is an error then. 
The function is supposed to return 1 byte only. If we wanted to return up to two bytes we would have to work on three bytes inside the function.
I propose to require that `length` is smaller than a byte (change the require to <8).

Reply:
Yes, then we'd need to change the conditional of https://github.com/clearmatics/zeth/blob/packing/zeth-contracts/contracts/BaseMixer.sol#L145-L149

Reply:
Updated.
The comment here is not the most clear I think: https://github.com/clearmatics/zeth/blob/aaa80b9d15d68f7c9b60ce60f6c5081beafe6003/zeth-contracts/contracts/BaseMixer.sol#L153-L154

We should say that we will work on at most field elements (not two bytes).
Then we should say that we (arbitrarily) assume that we extract at most one byte (hence the require). As 8 bits (a byte) can be written over at most two bytes, we will work at most on two bytes.

Reply:
> We should say that we will work on at most field elements (not two bytes)

> As 8 bits (a byte) can be written over at most two bytes, we will work on at most two bytes.

Yes happy to have this changed to say that we're manipulating at most 2 field elements, and that 8 consecutive bits are (at most) written on 2 bytes, if you think it is clearer @rrtoledo

Reply:
Updated.
I also added a couple of require on the other inputs.

Unfortunately this led to a `stack too deep`, so I had to remove a couple of variables (but added comments to compensate).